apps/aecore/lib/aecore/channel/worker.ex

```diff
@@ -0,0 +1,532 @@
+  def import_channel(channel_state) do
+    id = ChannelStatePeer.id(channel_state)
+    GenServer.call(__MODULE__, {:import_channel, id, channel_state})
+  end
+
+  @doc """
+  Import channel from open tx. Assumes no transactions were made
+  """
+  @spec import_from_open(SignedTx.t(), non_neg_integer(), role()) :: :ok | error()
+  def import_from_open(open_tx, reserve, role) do
+    peer_state = ChannelStatePeer.from_open(open_tx, reserve, role)
+    import_channel(peer_state)
+  end
+
+  @doc """
+  Imports channel from open tx and ChannelStateOffChain.
+  """
+  @spec import_from_open_and_state(
+          SignedTx.t(),
+          ChannelStateOffChain.t(),
+          non_neg_integer(),
+          role()
+        ) :: :ok | error()
+  def import_from_open_and_state(open_tx, state, reserve, role) do
+    peer_state = ChannelStatePeer.from_open_and_state(open_tx, state, reserve, role)
+    import_channel(peer_state)
+  end
+
+  @doc """
+  Initializes channel with temporary ID. This has to be called for every channel by both :initiator and :responder.
+  """
+  @spec initialize(
+          binary(),
+          {{Wallet.pubkey(), non_neg_integer()}, {Wallet.pubkey(), non_neg_integer()}},
+          role(),
+          non_neg_integer()
+        ) :: :ok | error()
+  def initialize(temporary_id, parties, role, channel_reserve) do
+    GenServer.call(
+      __MODULE__,
+      {:initialize, temporary_id, parties, role, channel_reserve}
+    )
+  end
+
+  @doc """
+  Creates open transaction. Can only be called once per channel by :initiator. Returns pair: generated channelID, half signed SignedTx.
+  """
+  @spec open(
+          binary(),
+          non_neg_integer(),
+          non_neg_integer(),
+          non_neg_integer(),
+          Wallet.privkey()
+        ) :: {:ok, binary(), SignedTx.t()} | error()
+  def open(temporary_id, locktime, fee, nonce, priv_key) do
+    GenServer.call(__MODULE__, {:open, temporary_id, locktime, fee, nonce, priv_key})
+  end
+
+  @doc """
+  Signs open transaction. Can only be called once per channel by :responder. Returns fully signed SignedTx and adds it to Pool.
+  """
+  @spec sign_open(binary(), SignedTx.t(), Wallet.privkey()) ::
+          {:ok, binary(), SignedTx.t()} | error()
+  def sign_open(temporary_id, open_tx, priv_key) do
+    GenServer.call(__MODULE__, {:sign_open, temporary_id, open_tx, priv_key})
+  end
+
+  @doc """
+  Notifies Channels Manager about confirmed channel open tx. Called by block validation stack.
+  """
+  @spec opened(SignedTx.t()) :: :ok
+  def opened(open_tx) do
+    case GenServer.call(__MODULE__, {:opened, open_tx}) do
+      :ok ->
+        :ok
+
+      {:error, reason} = error ->
+        Logger.warn(reason)
+        error
+    end
+  end
+
+  @doc """
+  Transfers amount to other peer in channel. Returns half-signed channel off-chain state. Can only be called on open channel.
+  """
+  @spec transfer(binary(), non_neg_integer(), Wallet.privkey()) ::
```

Comment:
Considering the [normal channel operation](https://github.com/aeternity/elixir-node/blob/GH-294-channels/README.md)
As far as I understand from the implementation of channels. Its ability is to transfer funds.
Question:
Is not it possible to extend the abstractness a bit, and instead of transferring only funds, can we pass on any kind of goods?
The GOODS could be any kind of data structure or may be a function. Who knows?
And the call would look like Channel.transfer(channel_id, GOODS, priv_key)
Of course, this kind of abstraction would lead to the need for writing appropriate handler.

Or I have a misconception of what the channels are, how they fit in blockchain and what their purpose is.

Reply:
State Channels can't interact with OnChain contracts AFAIK and those are used to store information about other goods. In future channels will be able to contain contracts, but I don't think they will be able to interact with on-chain contracts (they might be able to read, but probably not write).